### PR TITLE
[CIT-552] mqtt port change test

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes
This PR captures the code changes to fix the mqtt port change system test.

Test starts the subscriber more reliable way instead of waiting using sleep.
Also checks if the subscriber logged messages successfully or not.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py [hunk 1]
```diff
@@ -93,6 +100,34 @@ def validate_tedge_mqtt(self):
...
+                return;
+            else:
+                time.sleep(1)
+                n += 1
+        self.assertFalse(True, abortOnError=True, assertMessage=None)        
+        
+    # Starting subscriber takes sometime, so instead of sleeping before moving
+    # to next operation, its better to query active connections
+    def start_subscriber(self):
+        num_clients_b4 = subprocess.getoutput("mosquitto_sub -p 8889 -t '$SYS/broker/clients/active' -C 1")
```
Comment (didier-wenzek) [inline tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py:118]: Info: `tedge mqtt sub '$SYS/broker/clients/active'` is working fine ;-)

Comment (PradeepKiruvale) [reply tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py:118]: It works only with your changes to the tedge, not with the old one!.

File: tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py [hunk 2]
```diff
@@ -66,13 +77,7 @@ def validate(self):
...
 
     def validate_tedge_mqtt(self):
-        # subscribe for messages
-        mqtt_sub = self.startProcess(
-            command=self.sudo,
-            arguments=[self.tedge, "mqtt", "sub", "tedge/measurements"],
-            stdouterr="mqtt_sub",
-            background=True,
-        )
+        self.start_subscriber()
```
Comment (abelikt) [inline tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py:80]: Is it possible that we move this to the setup function? We have a four phase test pattern but here we are still starting thing in the validation part.

File: tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py (hunk 1 above)
Comment (abelikt) [reply tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py:118]: Good trick, I haven't thought of this. On the drawback side, now we need to wait for 10s. The process is probably up and running far earlier :)

**Comment by abelikt** [comment]
This test is now really getting complex. We also seem to loose structure a bit as setup, exercise and validation parts are a bit mixed. Can you also please update the test description. That is now really outdated

**Review by didier-wenzek — COMMENTED** [review]
This test seems a bit to complex for the task.

Why not start a `sub` commands to register all the messages published on `tedge/#`,
then send a message to the collectd's topic
to finally check that the expected reaction-messages has been published in cascade by the collectd mapper and the c8y mapper?

File: tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py (hunk 1 above)
Comment (PradeepKiruvale) [reply tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py:118]: Right, we have to wait for the mosquito statistics to get updated. Its not instant, it takes some time.

**Comment by PradeepKiruvale** [comment]
> This test seems a bit to complex for the task.
> 
> Why not start a `sub` commands to register all the messages published on `tedge/#`,
> then send a message to the collectd's topic
> to finally check that the expected reaction-messages has been published in cascade by the collectd mapper and the c8y mapper?

Collectd does not work with port change because the collectd.conf has to be updated with the newer port, which is a manual step.

File: tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py (hunk 2 above)
Comment (PradeepKiruvale) [reply tests/PySys/mqtt_port/mqtt_port_change_connection_works/run.py]: I completely removed this function. I feel the issue was about logging the message not about starting the subscriber process itself.

**Comment by PradeepKiruvale** [comment]
> This test is now really getting complex. We also seem to loose structure a bit as setup, exercise and validation parts are a bit mixed. Can you also please update the test description. That is now really outdated

Updated, please check.

**Review by abelikt — APPROVED** [review]
Looks good for me. Nevertheless, we should try and introduce the proposal from Didier soon: Sketch a tool to record MQTT messages #422

**Comment by PradeepKiruvale** [comment]
> Looks good for me. Nevertheless, we should try and introduce the proposal from Didier soon: Sketch a tool to record MQTT messages #422

Sure, we use this once we have this is ready.

**Pull request merged**